File: objc/macapp/WindowController.m
//
//  WindowController.m
//  macapp
//
//  Created by v on 2020/11/17.
//

#import "WindowController.h"

@interface WindowController () <NSWindowDelegate>

@end

@implementation WindowController

- (instancetype)initWithContentSize:(NSSize)size {
    NSWindowStyleMask styleMask = NSWindowStyleMaskClosable | NSWindowStyleMaskMiniaturizable | NSWindowStyleMaskResizable |
                                    NSWindowStyleMaskTitled | NSWindowStyleMaskUnifiedTitleAndToolbar | NSWindowStyleMaskFullSizeContentView;

    NSWindow *window = [[NSWindow alloc] initWithContentRect:NSMakeRect(0, 0, size.width, size.height)
                                                   styleMask:styleMask
                                                     backing:NSBackingStoreBuffered
                                                       defer:YES];

    window.titlebarAppearsTransparent = YES;

    if (self = [super initWithWindow:window]) {
        self.window.delegate = self;
    }

    return self;
}

- (void)windowDidLoad {
    [super windowDidLoad];
    
    // Implement this method to handle any initialization after your window controller's window has been loaded from its nib file.
}

- (void)setContentViewController:(NSViewController *)contentViewController {
    NSRect frame = self.window.frame;
    frame.origin = NSZeroPoint;
    contentViewController.view.frame = frame;
    super.contentViewController = contentViewController;
}

- (void)showWindow:(id)sender {
    [self.window center];
    [super showWindow:sender];
}

@end

File: swift/macapp/WindowController.swift
//
//  WindowController.swift
//  macapp
//
//  Created by v on 2020/11/17.
//

import Cocoa

class WindowController: NSWindowController, NSWindowDelegate {

    override var contentViewController: NSViewController? {
        willSet {
            if let newValue = newValue, let window = window {
                var frame = window.frame
                frame.origin = .zero
                newValue.view.frame = frame
            }
        }
    }

    convenience init(contentSize size: NSSize) {
        let styleMask: NSWindow.StyleMask = [.closable, .miniaturizable, .resizable,
                                             .titled, .unifiedTitleAndToolbar, .fullSizeContentView]

        let frame = NSRect(x: 0, y: 0, width: size.width, height: size.height)
        let window = NSWindow(contentRect: frame,
                              styleMask: styleMask,
                              backing: .buffered,
                              defer: true)
        
        window.titlebarAppearsTransparent = true

        self.init(window: window)

        window.delegate = self
    }

    override func showWindow(_ sender: Any?) {
        window?.center()
        super.showWindow(sender)
    }
}
